api: revert the private_key_provider_list and add fallback field to ProviderKeyProvider api

soulxu opened the pull request:
Commit Message: api: move private_key_provider_list to a new contrib extension
Additional Description:
According to the discussion in https://github.com/envoyproxy/envoy/pull/29004#issuecomment-1685751282, consider to implement the private_key_provider_list in an extensions.

This PR removes the `private_key_provider_list` from the `TlsCertificate` (was added at https://github.com/envoyproxy/envoy/pull/28215). Adds an new contrib extension `composite` for the private key provider. It will accept the a list of private key providers just like previous proposal.

Also add `fallback` field PrivateKeyProvider message to control the fallback behavior.

Risk Level: low
Testing: n/a
Docs Changes: API doc
Release Notes: n/a
Part of Fixes https://github.com/envoyproxy/envoy/issues/28214
Related to #29004

soulxu commented:
cc @wbpcode @ggreenway @markdroth

wbpcode commented on api/envoy/extensions/transport_sockets/tls/v3/common.proto:
```diff
@@ -178,23 +178,15 @@ message PrivateKeyProvider {
+  // If the private key provider isn't available (eg. the required hardware capability doesn't existed),
+  // Envoy will fallback to the BoringSSL default implementation when the `fallback` is true.
+  // The default value is `false`.
+  bool fallback = 4;
```
nit and just for discussion: would `is_optional` be better? I think there are lots of extension points have similar behavior and `is_optional` is widely used to declare a extension is optional and if the extension is not supported, some default behavior will be enabled.

soulxu replied:
In other place(like in the optional http filter), when we use the `is_optional`, it means it is fine if the extension isn't existed.

But at here, I suppose we only fallback when the provider can't be initialized correctly (like the required hardware capabilities not existed on the host). If the provider isn't existed, I still expect we will reject the config.

wbpcode replied:
IMO, it should be OK to use this new field to tell envoy to fallback to default behavior when the provider cannot work (either because hardware capabilities or provider isn't existed).

But your response sounds also reasonable to me. I have no strong point to this. cc @envoyproxy/api-shepherds

mattklein123 replied:
This seems fine to me.

wbpcode reviewed (commented):
LGTM overall. Thanks for all your patience and hard work to this area. 🌹

wbpcode commented:
We introduced this API in the #28215. And after discussion in the https://github.com/envoyproxy/envoy/pull/29004#issuecomment-1685751282, we think this API change is unnecessary and will bring lots of noise to our exist users. So, we incline to revert it.

It was marked as `unimplemented-hide`, and believe the grpc havn't use it, so it should be safe and acceptable to revert it.
cc @envoyproxy/api-shepherds

soulxu commented:
> We introduced this API in the #28215. And after discussion in the [#29004 (comment)](https://github.com/envoyproxy/envoy/pull/29004#issuecomment-1685751282), we think this API change is unnecessary and will bring lots of noise to our exist users. So, we incline to revert it.
> 
> It was marked as `unimplemented-hide`, and believe the grpc havn't use it, so it should be safe and acceptable to revert it. cc @envoyproxy/api-shepherds

If we agree on this, then later we will have a separated PR for a new extension which API looks like:

```
message FallbackProviderMethodConfig {
  // A list of private key providers, and at least one private key provider provided.
  repeated PrivateKeyProvider private_key_provider = 1 [(validate.rules).repeated = {min_items: 1}];
}
```

wbpcode commented:
Re-assgin this to @mattklein123 for API review because @adisuissa will be OOO in next days.

soulxu commented:
> > We introduced this API in the #28215. And after discussion in the [#29004 (comment)](https://github.com/envoyproxy/envoy/pull/29004#issuecomment-1685751282), we think this API change is unnecessary and will bring lots of noise to our exist users. So, we incline to revert it.
> > It was marked as `unimplemented-hide`, and believe the grpc havn't use it, so it should be safe and acceptable to revert it. cc @envoyproxy/api-shepherds
> 
> If we agree on this, then later we will have a separated PR for a new extension which API looks like:
> 
> ```
> message FallbackProviderMethodConfig {
>   // A list of private key providers, and at least one private key provider provided.
>   repeated PrivateKeyProvider private_key_provider = 1 [(validate.rules).repeated = {min_items: 1}];
> }
> ```

I added the extension API in this PR. But let me know if we want to propose those thing separately. The whole think can be separated as three part.

* Add `fallback` field to PrivateKeyProvider message and enable the fallback behavior in Envoy core
* Deprecate the `private_key` field in each provider, enable each provide reference the private key in the TlsCertificate::private_key
* Implement the new contrib extension `composite` provider

wbpcode commented:
At least I think the extension API should be moved to a seperated PR.

soulxu commented:
> At least I think the contrib extension API should be moved to a seperated PR.

got it, thanks
